### ISSUE TITLE
virtual-dom-rs bug: "Node's should only receive change text patches"

### ISSUE THREAD

**Issue by qthree** [chinedufn/percy#39]
Either diff or patch in virtual-dom-rs has bug. Probably something related to node indexing. It panics on tag&text change. Here is code (added to isomorphic example home view), that causes panic:
```rust
<table>
{
    (0..click_count.parse().unwrap()).into_iter().rev().map(|i| {
        if i%2 == 0 {
            html!{
                <tr><th>{i.to_string()}</th></tr>
            }
        } else {
            html!{
                <tr><td>{i.to_string()}</td></tr>
            }
        }
        
    }).collect::<Vec<_>>()
}
</table>
```
I also made fork with this code, for convenience: https://github.com/qthree/percy/tree/patch_bug
With this standard example if you open http://localhost:7878/?init=2 and start clicking "Click me", it will insert tags to the wrong parent. And if open http://localhost:7878/?init=4 then it will panic on the first click.

**Comment by qthree** [comment]
Looks like vdom state stays correct, but `patch::find_nodes` has indexing issue. For such html:
```html
<table>
  <tr>
    <td>1</td>
  </tr>
  <tr>
    <th>0</th>
  </tr>
</table>
```
With such patch:
```rust
Replace(19, VirtualNode | tag: th, props: {}, text: None, children: Some(
    [
        VirtualNode | tag: , props: {}, text: Some(
            "2"
        ), children: Some(
            []
        ) |
    ]
) |), Replace(21, VirtualNode | tag: td, props: {}, text: None, children: Some(
    [
        VirtualNode | tag: , props: {}, text: Some(
            "1"
        ), children: Some(
            []
        ) |
    ]
) |)
```
`patch::find_nodes` places in `elements_to_patch` wrong tags:
`19: <td>1</td>` - Ok
`21: <tr><th>0</th></tr>` - Wrong, should be `<th>0</th>`

So it replaces parental tag instead. And the further it goes the worse indexing shift gets.

**Comment by qthree** [comment]
Well, it looks that `patch::find_nodes` counts text node, but vdom not (all). Or i can't understand something.
![screenshot_20181108_130808](https://user-images.githubusercontent.com/9898555/48180726-599e3300-e357-11e8-95b2-1df211af5a9e.png)

**Comment by qthree** [comment]
I think i found the problem: `virtual_dom_rs::diff::diff_recursive` use early return when node's tag is changed, but because of that it doesn't count `cur_node_idx` for this node's children:
```rust
if old.tag != new.tag {
    patches.push(Patch::Replace(*cur_node_idx, &new));
    return patches;
}
...
for index in 0..min(old_child_count, new_child_count) {
    *cur_node_idx = *cur_node_idx + 1;
...
```

**Comment by qthree** [comment]
Also, when `new_child_count < old_child_count` old children are just dropped, so they and their nested children are not counted.

**Comment by chinedufn** [comment]
Fixed in version 0.3.1
